phpBB/phpbb/template/context.php

```diff
@@ -179,70 +179,65 @@ public function &get_root_ref()
-				$str = &$str[$blocks[$i]];
-				$str = &$str[sizeof($str) - 1];
-			}
 
-			$s_row_count = isset($str[$blocks[$blockcount]]) ? sizeof($str[$blocks[$blockcount]]) : 0;
-			$vararray['S_ROW_COUNT'] = $vararray['S_ROW_NUM'] = $s_row_count;
+		// For nested block, $blockcount > 0, for top-level block, $blockcount == 0
+		$blocks = explode('.', $blockname);
+		$blockcount = sizeof($blocks) - 1;
 
-			// Assign S_FIRST_ROW
-			if (!$s_row_count)
+		$block = &$this->tpldata;
+		for ($i = 0; $i < $blockcount; $i++)
+		{
+			if (($pos = strpos($blocks[$i], '[')) !== false)
 			{
-				$vararray['S_FIRST_ROW'] = true;
-			}
+				$name = substr($blocks[$i], 0, $pos);
 
-			// Assign S_BLOCK_NAME
-			$vararray['S_BLOCK_NAME'] = $blocks[$blockcount];
-
-			// Now the tricky part, we always assign S_LAST_ROW and remove the entry before
-			// This is much more clever than going through the complete template data on display (phew)
-			$vararray['S_LAST_ROW'] = true;
-			if ($s_row_count > 0)
+				if (strpos($blocks[$i], '[]') === $pos)
+				{
+					$index = sizeof($block[$name]) - 1;
+				}
+				else
+				{
+					$index = min((int) substr($blocks[$i], $pos + 1, -1), sizeof($block[$name]) - 1);
+				}
+			}
+			else
 			{
-				unset($str[$blocks[$blockcount]][($s_row_count - 1)]['S_LAST_ROW']);
+				$name = $blocks[$i];
+				$index = sizeof($block[$name]) - 1;
 			}
-
-			// Now we add the block that we're actually assigning to.
-			// We're adding a new iteration to this block with the given
-			// variable assignments.
-			$str[$blocks[$blockcount]][] = $vararray;
+			$block = &$block[$name];
+			$block = &$block[$index];
 		}
-		else
-		{
-			// Top-level block.
-			$s_row_count = (isset($this->tpldata[$blockname])) ? sizeof($this->tpldata[$blockname]) : 0;
-			$vararray['S_ROW_COUNT'] = $vararray['S_ROW_NUM'] = $s_row_count;
 
-			// Assign S_FIRST_ROW
-			if (!$s_row_count)
-			{
-				$vararray['S_FIRST_ROW'] = true;
-			}
+		// $block = &$block[$blocks[$i]]; // Do not traverse the last block as it might be empty
+		$name = $blocks[$i];
 
-			// Assign S_BLOCK_NAME
-			$vararray['S_BLOCK_NAME'] = $blockname;
+		// Assign S_ROW_COUNT and S_ROW_NUM
+		$s_row_count = isset($block[$name]) ? sizeof($block[$name]) : 0;
+		$vararray['S_ROW_COUNT'] = $vararray['S_ROW_NUM'] = $s_row_count;
 
-			// We always assign S_LAST_ROW and remove the entry before
-			$vararray['S_LAST_ROW'] = true;
-			if ($s_row_count > 0)
-			{
-				unset($this->tpldata[$blockname][($s_row_count - 1)]['S_LAST_ROW']);
-			}
+		// Assign S_FIRST_ROW
+		if (!$s_row_count)
+		{
+			$vararray['S_FIRST_ROW'] = true;
+		}
+
+		// Assign S_BLOCK_NAME
+		$vararray['S_BLOCK_NAME'] = $name;
 
-			// Add a new iteration to this block with the variable assignments we were given.
-			$this->tpldata[$blockname][] = $vararray;
+		// Now the tricky part, we always assign S_LAST_ROW and remove the entry before
+		// This is much more clever than going through the complete template data on display (phew)
```

Comment:
I don't know, but I guess so: this os the original unchanged code and comment...